Tests/t_ldapobject.py

```diff
@@ -643,6 +646,61 @@ def test105_reconnect_restore(self):
             self.server._start_slapd()
         self.assertEqual(l1.whoami_s(), 'dn:'+bind_dn)
 
+    def test_106_reconnect_restore(self):
+        lo = self.ldap_object_class(self.server.ldap_uri, retry_max=2, retry_delay=1)
+        bind_dn = 'cn=user1,' + self.server.suffix
+        lo.simple_bind_s(bind_dn, 'user1_pw')
+
+        dn = lo.whoami_s()[3:]
+
+        self.server._proc.terminate()
+        self.server.wait()
+
+        # do a search, wait for the timeout, handle SERVER_DOWN exception
+        with self.assertRaises(ldap.SERVER_DOWN):
+            print(lo.search_s(dn, ldap.SCOPE_BASE, '(objectClass=*)'))
+
+        self.server._start_slapd()
+
+        # try to use the connection again
+        print(lo.search_s(dn, ldap.SCOPE_BASE, '(objectClass=*)'))
+        # this used to raise ldap.INSUFFICIENT_ACCESS, as it is became a unbound connection and did everything anonymously.
+
+    def test_107_reconnect_restore(self):
+        excs = []
```

Comment:
Could you put a docstring here describing the intent behind this test so that someone can reimplement the busy loop of 100 threads with more intelligent thread coordination/asyncio in the future?

Reply:
@mistotebe OK, added a docstring. Can you check if suitable?